### PR TITLE
Ensure that the compiler only panics for legitimate bugs

### LINKED ISSUE #436: Remove user-facing `todo`s and `unimplemented`s
### What is wrong?

The compiler will crash during analysis of features that are not yet implemented. For this reason, we have been filtering out crashes that display the text "not implemented" or "not yet implemented" when fuzzing. This would be fine if we didn't also use these panic messages in later stages of compilation, but we are, and it's likely that we've been filtering out meaningful crashes.

### How can it be fixed

- replace all justifiable `todo`s and `unimplemented`s with graceful analyzer errors
- let agroce know that any panic is considered a bug

### PR DESCRIPTION
**Pull request by g-r-a-n-t**
### What was wrong?

closes #436

### How was it fixed?

Refactored front-end "not implemented" errors into analyzer errors and removed questionable variants.

### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [ ] OPTIONAL: Update [Spec](https://github.com/ethereum/fe/blob/master/docs/src/spec/index.md) if applicable
- [x] Add entry to the [release notes](https://github.com/ethereum/fe/blob/master/newsfragments/README.md) (may forgo for trivial changes)

- [x] Clean up commit history

### CODE REVIEW

File: analyzer/src/traversal/expressions.rs [hunk 1]
```diff
@@ -983,7 +996,7 @@ fn expr_call_value_attribute(
             ValueMethod::AbiEncode => match &value_attributes.typ {
                 Type::Struct(struct_) => {
                     if value_attributes.final_location() != Location::Memory {
```
Comment (g-r-a-n-t) [inline analyzer/src/traversal/expressions.rs:998]: these should just be "not in memory errors"

File: analyzer/src/traversal/expressions.rs [hunk 2]
```diff
@@ -1007,9 +1020,21 @@ fn expr_call_value_attribute(
                         Location::Memory,
                     ))
                 }
-                _ => todo!(),
+                _ => {
+                    context.not_yet_implemented("fix this error", func.span);
```
Comment (g-r-a-n-t) [inline analyzer/src/traversal/expressions.rs:1024]: come up with a good message

File: analyzer/src/traversal/expressions.rs [hunk 3]
```diff
@@ -1007,9 +1020,21 @@ fn expr_call_value_attribute(
...
+                    Ok(ExpressionAttributes::new(
+                        Type::unit(),
+                        Location::Value,
+                    ))
+                },
+            },
+            ValueMethod::AbiEncodePacked => {
```
Comment (g-r-a-n-t) [inline analyzer/src/traversal/expressions.rs:1031]: delete this variant

File: analyzer/src/traversal/expressions.rs [hunk 4]
```diff
@@ -556,7 +556,20 @@ fn expr_unary_operation(
...
             }
-            _ => todo!(),
+            UnaryOperator::Invert => {
+                context.not_yet_implemented("unary invert", exp.span);
+                Ok(ExpressionAttributes::new(
+                    Type::unit(),
+                    Location::Value,
+                ))
+            },
+            UnaryOperator::UAdd => {
```
Comment (g-r-a-n-t) [inline analyzer/src/traversal/expressions.rs:566]: what is this supposed to do?

File: analyzer/src/traversal/contracts.rs [hunk 5]
```diff
@@ -126,10 +126,11 @@ fn event_def(
         }
 
         // check if they are trying to index an array type
+        // todo clean all this up
         for index in indexed_fields.clone() {
             match fields[index].1.to_owned() {
                 FixedSize::Base(_) => {}
-                _ => unimplemented!("non-base type indexed event params"),
+                _ => context.not_yet_implemented("non-base type indexed event parameters", stmt.span),
```
Comment (g-r-a-n-t) [inline analyzer/src/traversal/contracts.rs:133]: fields

File: analyzer/src/traversal/structs.rs [hunk 6]
```diff
@@ -20,7 +20,7 @@ pub fn struct_def(
         if let Type::Base(base_typ) = field_type {
             val.add_field(&name.kind, &FixedSize::Base(base_typ))?;
         } else {
-            todo!("Non-Base type fields aren't yet supported")
+            context.not_yet_implemented("non-base type struct parameters", field.span)
```
Comment (g-r-a-n-t) [inline analyzer/src/traversal/structs.rs:23]: fields

File: analyzer/src/traversal/expressions.rs (hunk 4 above)
Comment (g-r-a-n-t) [reply analyzer/src/traversal/expressions.rs:566]: as far as I can tell, this is sometimes used to convert things to numbers.

is there another use case that Im missing? if not, I think we should just ditch it.

File: analyzer/src/traversal/expressions.rs [hunk 7]
```diff
@@ -983,7 +990,14 @@ fn expr_call_value_attribute(
...
                     if value_attributes.final_location() != Location::Memory {
-                        todo!("encode structs from storage")
+                        context.fancy_error(
+                            "value must be copied to memory",
+                            vec![Label::primary(value.span, "this value is in storage")],
+                            vec!["Hint: values located in storage can be copied to memory using the `to_mem` function.".into(),
+                                 "Example: `self.my_array.to_mem().abi_encode()`".into(),
+                            ],
+                        );
+                        // value_attributes.move_location = Some(Location::Memory);
```
Comment (g-r-a-n-t) [inline analyzer/src/traversal/expressions.rs:1000]: fix this

File: analyzer/src/traversal/module.rs [hunk 8]
```diff
@@ -28,8 +28,12 @@ pub fn module(context: &mut Context, module: &fe::Module) -> Result<(), Semantic
...
                 contracts.push((stmt, contract_scope))
             }
-            fe::ModuleStmt::FromImport { .. } => unimplemented!(),
-            fe::ModuleStmt::SimpleImport { .. } => unimplemented!(),
+            fe::ModuleStmt::FromImport { .. } => {
+                context.not_yet_implemented("from import", stmt.span)
+            }
+            fe::ModuleStmt::SimpleImport { .. } => {
+                context.not_yet_implemented("simple import", stmt.span)
+            }
```
Comment (sbillig) [inline analyzer/src/traversal/module.rs:36]: We could probably cut these from the ast too (though I have no strong feelings about it)

**Review by sbillig — APPROVED** [review]

**Review by satyamakgec — APPROVED** [review]

File: analyzer/src/traversal/module.rs (hunk 8 above)
Comment (g-r-a-n-t) [reply analyzer/src/traversal/module.rs:36]: Decided I'd leave it in there, since we may still use this syntax for imports.

**Pull request merged**